saga.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2016 - Max Ekman <max@looplab.se>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventhorizon
+
+import (
+	"log"
+)
+
+// Saga is an interface for a CQRS saga that listens to events and generate
+// commands. It is used for any long lived transaction and can be used to react
+// on multiple events.
+type Saga interface {
+	// SagaType returns the type of the saga.
+	SagaType() SagaType
+
+	// RunSaga handles an event in the saga that can return commands.
+	RunSaga(event Event) []Command
+}
+
+// SagaType is the type of a saga, used as its unique identifier.
+type SagaType string
+
+// SagaBase is a CQRS saga base to embed in domain specific sagas.
+//
+// A typical saga example:
+//   type OrderSaga struct {
+//       *eventhorizon.SagaBase
+//
+//       amount int
+//   }
+//
+// The implementing saga must set itself as the saga in the saga base.
+type SagaBase struct {
+	saga       Saga
+	commandBus CommandBus
```

Comment:
Yes, on an interface level it would be enough, but I think it better communicates the intent (that a saga can send commands to many aggregates) by having it use the command bus. Thoughts on that semantic?